rmw_connext_shared_cpp/src/shared_functions.cpp

```diff
@@ -1053,32 +1296,38 @@ get_node_names(const char * implementation_identifier,
     return RMW_RET_ERROR;
   }
   auto length = handles.length() + 1;  // add yourself
-  node_names->size = length;
-  node_names->data = static_cast<char **>(rmw_allocate(length * sizeof(char *)));
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rcutils_ret_t rcutils_ret = rcutils_string_array_init(node_names, length, &allocator);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string_safe())
+    return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+  }
 
   DDS_DomainParticipantQos participant_qos;
   DDS_ReturnCode_t status = participant->get_qos(participant_qos);
   if (status != DDS_RETCODE_OK) {
     RMW_SET_ERROR_MSG("failed to get default participant qos");
     return RMW_RET_ERROR;
   }
-  auto participant_name_length = strlen(participant_qos.participant_name.name) + 1;
-  node_names->data[0] =
-    static_cast<char *>(rmw_allocate(participant_name_length * sizeof(char)));
-  snprintf(node_names->data[0], participant_name_length, "%s",
-    participant_qos.participant_name.name);
+  node_names->data[0] = rcutils_strdup(participant_qos.participant_name.name, allocator);
+  if (!node_names->data[0]) {
+    RMW_SET_ERROR_MSG("could not allocate memory for node name")
+    return RMW_RET_BAD_ALLOC;
+  }
 
 
   for (auto i = 1; i < length; ++i) {
     ParticipantBuiltinTopicData pbtd;
     auto dds_ret = participant->get_discovered_participant_data(pbtd, handles[i - 1]);
-    char * name = pbtd.participant_name.name;
+    const char * name = pbtd.participant_name.name;
     if (!name || dds_ret != DDS_RETCODE_OK) {
-      name = const_cast<char *>("(no name)");
+      name = "(no name)";
+    }
+    node_names->data[i] = rcutils_strdup(name, allocator);
+    if (!node_names->data[i]) {
+      RMW_SET_ERROR_MSG("could not allocate memory for node name")
+      return RMW_RET_BAD_ALLOC;
```

Comment:
Needs to deallocate previously allocated `data[i]`. Same as for `rmw_fastrtps_cpp`.